Add support for Markdown templates

Ikuni17 opened the pull request:
## Summary
Closes elastic/kibana-operations#283

- Adds support for Markdown templates and generates Markdown release notes for Serverless and `v9+` of other solutions.
- Removes Security label restrictions on Serverless release notes
- Deprecations and Breaking changes have their own files in the docs system, when we generate here it's all the same file but they are separated at the end.

## Testing
0. Pull repo, `yarn install`, checkout `feature/283/markdown-support`
1. `yarn start`
2. Generate notes!

## Examples
<details>
<summary><h3>Serverless</h3></summary>
<br>

```md
## May 5, 2025 [serverless-changelog-05052025]

### Features and enhancements [serverless-changelog-05052025-features-enhancements]

* Fleet agents tag filter is searchable and sorted [#219639]({{kib-pull}}219639).
* ES|QL rule type improvements - adding grouping per row [#212135]({{kib-pull}}212135).
* Added compact view in monitors overview page [#219060]({{kib-pull}}219060).
* Investigation Guide backend [#216377]({{kib-pull}}216377).
* Adds context.grouping action variable in SLO Burn rate and ES Query rules [#213550]({{kib-pull}}213550).
* 790 deployment environment discrepancy [#217899]({{kib-pull}}217899).
* Adds Edot logging level to central config [#219722]({{kib-pull}}219722).
* Updates kibana MITRE data to `v16.1` [#215026]({{kib-pull}}215026).
* Fixes some OpenAI models not accepting temperature for Inference service [#218887]({{kib-pull}}218887).
* Update styles for Color formatter to look like a badge [#189391]({{kib-pull}}189391).
* Adds the ability to switch between relative and absolute time range [#218056]({{kib-pull}}218056).


### Fixes [serverless-changelog-05052025-fixes]

* Fixes ignored dynamic templates [#219875]({{kib-pull}}219875).
* [ES|QL] Sync the Dashboard ES|QL query and filters with the corresponding in Visualizations in *Lens* [#218997]({{kib-pull}}218997).
* Fixes option list control making 2 requests on refresh [#219625]({{kib-pull}}219625).
* Fixes capability required for Siem Migrations Topic [#219427]({{kib-pull}}219427).
* Monitor status rule - remove return to avoid limiting alerts [#219291]({{kib-pull}}219291).
% !!TODO!! The above PR had a lengthy release note description:
% Synthetics monitor status rule - Ensures that an individual alert is sent per monitor configuration when the "Receive distinct alerts per location" toggle is unchecked.
* Fixed monitor status rule control [#218994]({{kib-pull}}218994).
* Don't open in flyout mode on mount [#219420]({{kib-pull}}219420).
* Change the alerts query to include environment not defined value [#219228]({{kib-pull}}219228).
* Fixes Alerts environment query follow up [#219571]({{kib-pull}}219571).
* Handle missing `trace` in API response [#219512]({{kib-pull}}219512).
* Handle create anomaly detection jobs failures [#219364]({{kib-pull}}219364).
* Optional chaining to prevent undefined error in `custom_link_flyout.tsx` [#219668]({{kib-pull}}219668).
* Query alerts using the `alert.start` field and update alerts function API test to check alert information [#219651]({{kib-pull}}219651).
* Fixes scroll in error flyout [#218697]({{kib-pull}}218697).
* Adding code to check privileges for enabling the Run Engine button [#213054]({{kib-pull}}213054).
* Fixes rule import error message display [#218701]({{kib-pull}}218701).
* Fixes broken icons in integrations coming from Home plugin [#219206]({{kib-pull}}219206).
* Remove check for unused Connector role [#219358]({{kib-pull}}219358).
* Ensure ability to change provider without error [#219020]({{kib-pull}}219020).
* Correct quotes in ES|QL queries for function arguments [#217680]({{kib-pull}}217680).
```
</details>

<details>
<summary><h3>Kibana 9.1.0</h3></summary>
<br>

```md
% FEATURES, ENHANCEMENTS, FIXES
% Paste in index.md

## 9.1.0 [kibana-9.1.0-release-notes]

% ::::{NOTE}
% ::::


### Features and enhancements [kibana-9.1.0-features-enhancements]

**Alerting**:
* Support rrule for task scheduling [#217728]({{kib-pull}}217728).
* Publish new public APIs for the Maintenance Window [#216756]({{kib-pull}}216756).
* Alerts created within a Maintenance Windows trigger actions after the MW expires [#219797]({{kib-pull}}219797).
* Moves rule settings to a flyout instead of a modal [#216162]({{kib-pull}}216162).
* Allow users to delete a snooze schedule from a rule using schedule id [#213247]({{kib-pull}}213247).
* Allow users to create a snooze schedule for rule via API [#210584]({{kib-pull}}210584).

**Dashboards and Visualizations**:
* "Compare to" badge for Metric chart in *Lens* [#214811]({{kib-pull}}214811).
* Enable read only editor mode to inspect panel's configuration in *Lens* [#208554]({{kib-pull}}208554).
* [dashboard] Basic keyboard interaction [#208286]({{kib-pull}}208286).
* Adds ability to open links in *Vega* in new window [#216200]({{kib-pull}}216200).
* Inject / extract tag references [#214788]({{kib-pull}}214788).
* Globe projection [#212437]({{kib-pull}}212437).
* Keeps the chart configuration changes done by the user when changing the query whenever it is possible [#210780]({{kib-pull}}210780).

**Data ingestion and Fleet**:
* Enable feature flag `enableAutomaticAgentUpgrades` [#219932]({{kib-pull}}219932).
* Adds Edit ReadMe Functionality To Custom Integrations [#215259]({{kib-pull}}215259).
% !!TODO!! The above PR had a lengthy release note description:
% Adds edit functionality to custom integrations, allowing a user to edit the README file of a custom integration and save it to be persisted. Additionally, saving will automatically increment the version of the integration and update all associated policies.
* Adds ssl fields to agent binary source settings [#213211]({{kib-pull}}213211).
* Expose ssl options for ES and remote ES outputs in UI [#208745]({{kib-pull}}208745).
* Adds SSL options to fleet server hosts settings [#208091]({{kib-pull}}208091).
* Fleet agents tag filter is searchable and sorted [#219639]({{kib-pull}}219639).
* Callout breaking changes on integration upgrade [#217257]({{kib-pull}}217257).
* Adds msi installer command for fleet server and agents [#217217]({{kib-pull}}217217).
* Optional ssl for fleet logstash output [#216216]({{kib-pull}}216216).
* Format last activity value in fleet agent details view as datetime [#215531]({{kib-pull}}215531).
* Support `searchAfter` and PIT (point-in-time) parameters for get agents list API [#213486]({{kib-pull}}213486).
* Register custom integrations search provider [#213013]({{kib-pull}}213013).

**Discover**:
* Creates control by typing a questionmark [#216839]({{kib-pull}}216839).
* Implementing click action for Stacktrace and Degraded Fields on Discover [#214413]({{kib-pull}}214413).
* Added context aware logic for logs view in discover to show Load More… [#211176]({{kib-pull}}211176).
* Adds an ES|QL control option on the dashboard controls dropdown [#219495]({{kib-pull}}219495).
* Adds 'Copy value' button to field value cells [#218817]({{kib-pull}}218817).
* Enable suggestions for `CHANGE_POINT` command [#218100]({{kib-pull}}218100).
* Autocomplete for `STATS...WHERE` [#216379]({{kib-pull}}216379).
* Validation and autocomplete support for the `CHANGE_POINT` command [#216043]({{kib-pull}}216043).
* Highlights the code examples in our inline docs [#214915]({{kib-pull}}214915).
* Suggest triple quotes when the user selects the `KQL` / `QSTR` [#211457]({{kib-pull}}211457).
* Display a warning and a tooltip for the `_score` column in the grid [#211013]({{kib-pull}}211013).
* Allow command/ctrl click for "New" action in top nav [#210982]({{kib-pull}}210982).
* Apply compact Display options Popover layout [#210180]({{kib-pull}}210180).
* Allows the creation of dynamic aggregations controls for ES|QL charts [#210170]({{kib-pull}}210170).

**Elastic Observability solution**:
For the Elastic Observability 9.1.0 release information, refer to [Elastic Observability Solution Release Notes](docs-content://release-notes/elastic-observability/index.md).

**Elastic Security solution**:
For the Elastic Security 9.1.0 release information, refer to [Elastic Security Solution Release Notes](docs-content://release-notes/elastic-security/index.md).

**Kibana platform**:
* Favorite a dashboard from within [#201596]({{kib-pull}}201596).

**Kibana security**:
* Adds defaultSolution to spaces config [#218360]({{kib-pull}}218360).

**Machine Learning**:
* Moves job and trained model management features into Stack Management [#204290]({{kib-pull}}204290).
* Fixes some OpenAI models not accepting temperature for Inference service [#218887]({{kib-pull}}218887).
* Support local file path for `xpack.productDocBase.artifactRepositoryUrl` [#217046]({{kib-pull}}217046).
* Adds Voyage AI and DeepSeek icons [#216651]({{kib-pull}}216651).
* Anomalies table: Enhances display for anomaly time function values [#216142]({{kib-pull}}216142).
* Trained models: Replace download button by extending deploy action [#205699]({{kib-pull}}205699).

**Management**:
* Remove enablement check in `PUT /api/streams/{id}` for classic streams [#212289]({{kib-pull}}212289).
* Enable `/api/streams/{id}/_group` endpoints for GroupStreams [#210114]({{kib-pull}}210114).
* Update styles for Color formatter to look like a badge [#189391]({{kib-pull}}189391).

**Platform**:
* Added an option to User Settings that allows the Kibana interface to display in a high contrast mode [#216242]({{kib-pull}}216242).

**Sharing**:
* Adds the ability to switch between relative and absolute time range [#218056]({{kib-pull}}218056).
* Adds 'page reload' screen reader warning [#214822]({{kib-pull}}214822).


### Fixes [kibana-9.1.0-fixes]

**Alerting**:
* Fixes ignored dynamic templates [#219875]({{kib-pull}}219875).
* Fixes a regression that caused the cases actions to disappear from the detections engine alerts table bulk actions menu [#215111]({{kib-pull}}215111).

**Dashboards and Visualizations**:
* Allows to retrieve empty columns [#218085]({{kib-pull}}218085).
* Do not break when the table has no data in *Lens* [#217937]({{kib-pull}}217937).
* Fixes coloring/palette assignment on partition charts in *Lens* [#215426]({{kib-pull}}215426).
* Fixes invalid dashboard displayed as 404 instead of showing validation error [#211661]({{kib-pull}}211661).
* Adds `useCustomDragHandle` prop [#210463]({{kib-pull}}210463).
* Fixes presentation panel styles [#210113]({{kib-pull}}210113).
* Store assignments as raw/serialized values [#207957]({{kib-pull}}207957).
% !!TODO!! The above PR had a lengthy release note description:
% This PR fixes an issue ([#193080](https://github.com/elastic/kibana/issues/193080)) where custom ranges and multi-field values were not correctly colored based on selected color mapping configurations.

**Data ingestion and Fleet**:
* Fixes ssl config overridden from advanced yaml in full agent policy [#219902]({{kib-pull}}219902).
* Fixes capability required for Siem Migrations Topic [#219427]({{kib-pull}}219427).
* Make output and fleet server non-editable for agentless policies [#218905]({{kib-pull}}218905).
* Support integrations having secrets with multiple values [#216918]({{kib-pull}}216918).
* Adds remote cluster instructions for syncing integrations [#211997]({{kib-pull}}211997).
* Update install snippets to include all platforms [#210249]({{kib-pull}}210249).

**Discover**:
* Fixes rename wrong validation for asterisc in name [#219832]({{kib-pull}}219832).
* Make icon only presentational [#219696]({{kib-pull}}219696).
* Fixes the wrong source validation in case of unknown patterns [#218352]({{kib-pull}}218352).
* Fixes editor menus on safari [#218167]({{kib-pull}}218167).
* Fixes the broken tooltip suggestions descriptions [#218067]({{kib-pull}}218067).
* Adds items count to fields accordion title aria-label [#216993]({{kib-pull}}216993).
* Discover esc closes flyout when focus is on filter [#216630]({{kib-pull}}216630).
* Enables the timepicker if the time params are used with cast [#215820]({{kib-pull}}215820).
* Calculate the query for retrieving the values correctly [#214905]({{kib-pull}}214905).
* Make sure that the variables in the editor are always up to date [#214833]({{kib-pull}}214833).
* Fixes the visibility of the datepicker [#214728]({{kib-pull}}214728).
* Fixes warnings with escaped quotes [#213685]({{kib-pull}}213685).
* Fixes the wrong validation when a named param is used as function [#213355]({{kib-pull}}213355).
* Supports date_nanos in bucket [#213319]({{kib-pull}}213319).
* Exclude Elasticsearch metadata fields from Display in Content Column [#213255]({{kib-pull}}213255).
* Fixes the suggestion problem in where for multiline queries [#213240]({{kib-pull}}213240).
* Suggest triple quotes when the user selects the `KQL` / `QSTR` [#211457]({{kib-pull}}211457).

**Elastic Observability solution**:
For the Elastic Observability 9.1.0 release information, refer to [Elastic Observability Solution Release Notes](docs-content://release-notes/elastic-observability/index.md).

**Elastic Security solution**:
For the Elastic Security 9.1.0 release information, refer to [Elastic Security Solution Release Notes](docs-content://release-notes/elastic-security/index.md).

**Kibana platform**:
* Remove listing limit warning [#217945]({{kib-pull}}217945).
* Custom link colour option for top banner [#214241]({{kib-pull}}214241).
* Change in sortBy dropdown component option name wording [#206464]({{kib-pull}}206464).

**Machine Learning**:
* Change Point Detection: Fixes 'Use full data' button issue which could cause page to crash [#217291]({{kib-pull}}217291).
* Fixing permission to view ML nodes [#215503]({{kib-pull}}215503).
* Fixes chart in single metric anomaly detection wizard [#214837]({{kib-pull}}214837).
* Trained Models: Prevent multiple clicks in Delete Model dialog [#211580]({{kib-pull}}211580).
* Trained Models: Fixes further deployment of models after a failed deployment [#211459]({{kib-pull}}211459).
* AIOps Hiding saved query controls [#210556]({{kib-pull}}210556).
* Only allow .ndjson files when bulk importing to the knowledge base [#215433]({{kib-pull}}215433).
* Automatic Import now produces correct pipeline to parse CSV files with special characters in column names [#212513]({{kib-pull}}212513).
* Improve finding a function [#210437]({{kib-pull}}210437).

% BREAKING CHANGES
% Paste in breaking-changes.md

## 9.1.0 [kibana-9.1.0-breaking-changes]
$$$kibana-213916$$$
::::{dropdown} Now the fields / functions variables are being described with ?? in the query. The values variables use ? as before 
% **Details**<br> Description
% **Impact**<br> Impact of the breaking change.
% **Action**<br> Steps for mitigating impact.
View [#213916]({{kib-pull}}213916).
::::
```
</details>

florent-leborgne commented:
Thanks, I also had this (started with Kibana) but yours is probably more robust :D https://github.com/elastic/kibana-release-notes/pull/43

delanni commented on tsconfig.json:
```diff
@@ -25,6 +25,6 @@
   ],
   "exclude": [
-    "node_modules/@elastic/eui/src"
+    "node_modules/@elastic/eui/src",
```
nit: this might break a regular `JSON.parse` - if not needed, let's keep the file compatible
```suggestion
    "node_modules/@elastic/eui/src"
```

Ikuni17 replied:
Pushed cb46b97d027b3f420613b3bedafbe265b5adda9f to revert to `main`

florent-leborgne commented on src/config/templates/common.ts:
```diff
@@ -0,0 +1,381 @@
+        `
+
+{{#prs.breaking}}
+---
+navigation_title: "Breaking changes"
+mapped_pages:
+  - https://www.elastic.co/guide/en/${urlPath}/current/breaking-changes-summary.html
+---
+# ${escapedTemplateNameTag} breaking changes [${name}-breaking-changes]
+Breaking changes can impact your Elastic applications, potentially disrupting normal operations. Before you upgrade, carefully review the ${escapedTemplateNameTag} breaking changes and take the necessary steps to mitigate any issues. To learn how to upgrade, check [Upgrade](/deploy-manage/upgrade.md).
```
Since we're always going to be pasting into the same files cumulatively, no need to regenerate the start of the files each time. 

I've been using this trimmed down version for the past 2 releases (for Kibana) and it's fine if you want to take a look: https://github.com/elastic/kibana-release-notes/pull/43/files#diff-7a2f2afb039ab3adb1760e1583fcd43fcc0d529866eb48e4a07fa6c8c438949eR301-R305

The only thing that's not ideal in my template above is that enhancements are after fixes instead of grouped together properly, in case you have ideas around making this a unified list :D

Ikuni17 replied:
> Thanks, I also had this (started with Kibana) but yours is probably more robust :D https://github.com/elastic/kibana-release-notes/pull/43
>
> I've been using this trimmed down version for the past 2 releases (for Kibana) and it's fine if you want to take a look

Thanks for the review @florent-leborgne, I will bring your template changes into this PR, then we can close #43. This PR will still handle ASCIIDOC for `8.19` releases.

> Since we're always going to be pasting into the same files cumulatively, no need to regenerate the start of the files each time. IMO

To be clear, is this only for breaking changes and deprecations or for the main notes generation as well?

> The only thing that's not ideal in my template above is that enhancements are after features instead of grouped together properly, in case you have ideas around making this a unified list :D

My template is the same as yours, listing all features, then all enhancements. What is the proper way they are supposed to be grouped? Also, I see in your template that there are only enhancements and no features in patch releases. Is this always the case? It makes sense, but I feel like we should include the feature PR just in case.

Ikuni17 replied:
Pushed 0d9f3457ee449ee98f54b25afcc4e6ca17ba81ec and 11992bc58d1a71ae580f935032e74464d7244f0e to bring in your changes from #43

Ikuni17 replied:
I also pushed 0dfdccec10d2c412133219fde6f5cbb4e542729c since your template doesn't have the start of the file and I assume that is correct. Addresses:

> To be clear, is this only for breaking changes and deprecations or for the main notes generation as well?

We can easily revert this commit my assumption was incorrect.

florent-leborgne replied:
Hey @Ikuni17 thank you so much for your help with this.

> To be clear, is this only for breaking changes and deprecations or for the main notes generation as well?

It's for all files, correct. Thank you!

> What is the proper way they are supposed to be grouped?

Right I saw that we were doing the same thing actually after my comment. Previously (in asciidoc) we had:

```
=== Features
* Discover feature 1
* Discover feature 2
* Management feature 1
* Management feature 2

=== Enhancements
* Discover enhancement 1
* Discover enhancement 2

etc.
```

Now in Markdown, in addition to the format change and as you noticed, we want to have:

```
### Features and enhancements
* Discover feature 1
* Discover feature 2
* Discover enhancement 1
* Discover enhancement 2
* Management feature 1
* Management feature 2

etc.
```

Meaning that in what we deliver at then end, it makes more sense to have features and enhancement of the same category to be together, instead of having all features, then all enhancements. (Do I make sense? 😅). I was not sure how to resolve this with my limited skills, so I initially just put them one after the other and figured we'd edit the ordering manually. But just in case you can think of something... No worries if that makes it all too complex as well.

> Also, I see in your template that there are only enhancements and no features in patch releases. Is this always the case? It makes sense, but I feel like we should include the feature PR just in case.

I think you are correct here. I basically just edited the asciidoc template into markdown to reproduce the same, but I may have misread the patch one. If there can be features in patches, it's best if we can prepare the template for that possibility. Thanks!

Ikuni17 commented on src/config/templates/common.ts:
```diff
@@ -0,0 +1,381 @@
+% ### Fixes [${name}-next-fixes]
+% *
+
+## {{version}} [${name}-{{versionWithoutPeriods}}-release-notes]
```
180f5681deedc0ded73df4ced343446487d5450f

Ikuni17 commented:
@delanni I pushed da981bf9eac6c99a7cb4c4cfb331856f54a74f14 to hopefully address the readability. Let me know if you have other ideas.

I have also updated the examples after this round of feedback. (cc @florent-leborgne)

florent-leborgne commented:
I still want to review this but haven't had the time yet :) Just looking at the output I think there are a few tweaks to make to the serverless one especially about this extract:

In serverless release notes Security and Observability changes should be included (we don't have a specific page for security serverless or observability serverless)

```
For the Elastic Observability 04282025 release information, refer to [Elastic Observability Solution Release Notes](docs-content://release-notes/elastic-observability/index.md).
For the Elastic Security 04282025 release information, refer to [Elastic Security Solution Release Notes](docs-content://release-notes/elastic-security/index.md).
For the Elastic Observability 04282025 release information, refer to [Elastic Observability Solution Release Notes](docs-content://release-notes/elastic-observability/index.md).
For the Elastic Security 04282025 release information, refer to [Elastic Security Solution Release Notes](docs-content://release-notes/elastic-security/index.md).
```

Ikuni17 commented:
@florent-leborgne If there are any small template changes remaining feel free to push to this branch since you will know better than me. The examples have been updated again to reflect new changes. I fixed the grouping of features and enhancements with 2231c6540ee5a66e1d4f84244044517ace72fdc2

> In serverless release notes Security and Observability changes should be included (we don't have a specific page for security serverless or observability serverless)

Fixed with 2f85d25d4156efbb36f130a40e658d9c34f8f86c.
